### PR TITLE
Simply json schema validation error messages

### LINKED ISSUE #11129: Config validation errors are overly verbose
Example from Luke trying out project config:
```
error: could not get cloud url: could not load current project: could not validate '/Users/lukehoban/dd/ced2022demos/demo1/Pulumi.yaml': 21 errors occurred:
	* #/config/minClusterSize: oneOf failed
	* #/config/minClusterSize: expected string, but got object
	* #/config/minClusterSize: expected integer, but got object
	* #/config/minClusterSize: expected boolean, but got object
	* #/config/minClusterSize: doesn't validate with '/$defs/configTypeDeclaration'
	* #/config/minClusterSize/type: doesn't validate with '/$defs/simpleConfigType'
	* #/config/minClusterSize/type: value must be one of "string", "integer", "boolean", "array"
	* #/config/maxClusterSize: oneOf failed
	* #/config/maxClusterSize: expected string, but got object
	* #/config/maxClusterSize: expected integer, but got object
	* #/config/maxClusterSize: expected boolean, but got object
	* #/config/maxClusterSize: doesn't validate with '/$defs/configTypeDeclaration'
	* #/config/maxClusterSize/type: doesn't validate with '/$defs/simpleConfigType'
	* #/config/maxClusterSize/type: value must be one of "string", "integer", "boolean", "array"
	* #/config/desiredClusterSize: oneOf failed
	* #/config/desiredClusterSize: expected string, but got object
	* #/config/desiredClusterSize: expected integer, but got object
	* #/config/desiredClusterSize: expected boolean, but got object
	* #/config/desiredClusterSize: doesn't validate with '/$defs/configTypeDeclaration'
	* #/config/desiredClusterSize/type: doesn't validate with '/$defs/simpleConfigType'
	* #/config/desiredClusterSize/type: value must be one of "string", "integer", "boolean", "array"
```

Not sure what we can do better here because we're just using json schema to validate, but maybe there's something we can improve.

Extra oddness that this comes up under "could not get cloud url:", maybe we should be pre-validating project files somehow so it's more clear it's a Pulumi.yaml issue.

### PR DESCRIPTION
**Pull request by aq17**
Fixes #11129 
Simplifies overly verbose json validation error messages (example in ticket)

### CODE REVIEW

File: sdk/go/common/workspace/project.go [hunk 1]
```diff
@@ -358,16 +358,16 @@ func ValidateProject(raw interface{}) error {
 	var errs *multierror.Error
 	var appendError func(err *jsonschema.ValidationError)
 	appendError = func(err *jsonschema.ValidationError) {
-		if err.InstanceLocation != "" && err.Message != "" {
+		if err.InstanceLocation != "" && err.Message != "" && len(err.Causes) == 0 {
```
Comment (aq17) [inline sdk/go/common/workspace/project.go:361]: We can already limit noise by not printing an error's message when it contains nested errors, as the nested errors typically are the root cause, i.e.
```
error: could not get cloud url: could not load current project: could not validate '/Users/alexqiu/github.com/pulumi/test-296/Pulumi.yaml': 8 errors occurred:
        * #/config/bucketNamePrefix: oneOf failed
        * #/config/bucketNamePrefix: expected string, but got object
        * #/config/bucketNamePrefix: expected integer, but got object
        * #/config/bucketNamePrefix: expected boolean, but got object
        * #/config/bucketNamePrefix: expected array, but got object
        * #/config/bucketNamePrefix: doesn't validate with '/$defs/configTypeDeclaration'
        * #/config/bucketNamePrefix/type: doesn't validate with '/$defs/simpleConfigType'
        * #/config/bucketNamePrefix/type: value must be one of "string", "integer", "boolean", "array"
```
becomes
`5 errors occurred:
        * #/config/bucketNamePrefix: expected string, but got object
        * #/config/bucketNamePrefix: expected integer, but got object
        * #/config/bucketNamePrefix: expected boolean, but got object
        * #/config/bucketNamePrefix: expected array, but got object
        * #/config/bucketNamePrefix/type: value must be one of "string", "integer", "boolean", "array"`

File: sdk/go/common/workspace/project.go [hunk 2]
```diff
@@ -358,16 +358,16 @@ func ValidateProject(raw interface{}) error {
...
 			errorf := func(path, message string, args ...interface{}) error {
 				contract.Require(path != "", "path")
 				return fmt.Errorf("%s: %s", path, fmt.Sprintf(message, args...))
 			}
 
 			errs = multierror.Append(errs, errorf("#"+err.InstanceLocation, "%v", err.Message))
 		}
-		for _, err := range err.Causes {
-			appendError(err)
+		if len(err.Causes) != 0 {
```
Comment (aq17) [inline sdk/go/common/workspace/project.go:369]: Going off of @AaronFriel 's insight https://github.com/pulumi/pulumi/issues/11129, I do think that the last/ innermost error cause takes precedence, such as 
```
        * #/config/bucketNamePrefix: expected string, but got object
        * #/config/bucketNamePrefix: expected integer, but got object
        * #/config/bucketNamePrefix: expected boolean, but got object
        * #/config/bucketNamePrefix: expected array, but got object
        * #/config/bucketNamePrefix/type: value must be one of "string", "integer", "boolean", "array"
```
The first four errors are due to the last error.

File: sdk/go/common/workspace/project.go [hunk 3]
```diff
@@ -357,20 +357,29 @@ func ValidateProject(raw interface{}) error {
...
+		contract.Require(path != "", "path")
+		return fmt.Errorf("%s: %s", path, fmt.Sprintf(message, args...))
+	}
 	appendError = func(err *jsonschema.ValidationError) {
-		if err.InstanceLocation != "" && err.Message != "" {
-			errorf := func(path, message string, args ...interface{}) error {
-				contract.Require(path != "", "path")
-				return fmt.Errorf("%s: %s", path, fmt.Sprintf(message, args...))
+		if err.InstanceLocation != "" && err.Message != "" && len(err.Causes) == 0 {
+			if e, ok := locToErr[err.InstanceLocation]; ok {
```
Comment (aq17) [inline sdk/go/common/workspace/project.go:367]: Concatenating errors from the same location cleans it up a bit more
```
2 errors occurred:
        * #/config/bucketNamePrefix: expected string, but got object; expected integer, but got object; expected boolean, but got object; expected array, but got object
        * #/config/bucketNamePrefix/type: value must be one of "string", "integer", "boolean", "array"
        ```

Comment (iwahbe) [reply sdk/go/common/workspace/project.go:367]: This is a major improvement! Would it be possible to go one step further with the first error:
```
2 errors occurred:
        * #/config/bucketNamePrefix: expected a string, integer, boolean or array, but got object
        * #/config/bucketNamePrefix/type: value must be one of "string", "integer", "boolean", "array"
```

I imagine we can do a post-folding step where we combine all messages of the form `expected a \([a-z]+\), but got \([a-z]+\)`.

File: sdk/go/common/workspace/project_test.go [hunk 4]
```diff
@@ -900,10 +898,8 @@ func TestProjectLoadYAML(t *testing.T) {
 	_, err = loadProjectFromText(t, "name: project\nruntime: 4")
 	// These can vary in order, so contains not equals check
 	expected := []string{
-		"3 errors occurred:",
-		"* #/runtime: oneOf failed",
-		"* #/runtime: expected string, but got number",
-		"* #/runtime: expected object, but got number"}
+		"1 error occurred:",
+		"* #/runtime: expected string, but got number; expected object, but got number"}
```
Comment (Frassle) [inline sdk/go/common/workspace/project_test.go:902]: I'd like to see more tests for this before merging.
We're currently missing errors for various config "mistakes" and we should write and merge those tests first, and then rebase these changes and see how they compare.
I'm not confident just dropping all intermediate messages will always be clear.

**Review by Frassle — CHANGES_REQUESTED** [review]

File: sdk/go/common/workspace/project_test.go (hunk 4 above)
Comment (aq17) [reply sdk/go/common/workspace/project_test.go]: @Frassle could you give a few examples of the type of config mistakes so I can check against what we currently have?

Comment (Frassle) [reply sdk/go/common/workspace/project_test.go]: Sure stuff like having:
* "value" + one of the other schema options.
* type set to "array" but no "items" block
* type not set to array but having an items block
* type set to array but items not being a valid schema block

Non config related:
* setting "backend" to a string directly
* mis-spelling "default" (e.g. "dflt") in a template config section
* Setting the plugins.providers array as a map
* Setting runtime.options to a string

File: sdk/go/common/workspace/creds.go [hunk 5]
```diff
@@ -39,6 +39,8 @@ const PulumiCredentialsPathEnvVar = "PULUMI_CREDENTIALS_PATH"
 // used instead of the currently logged in backend or the current projects backend.
 const PulumiBackendURLEnvVar = "PULUMI_BACKEND_URL"
 
+const LoadCurrentProjectFailMsg = "could not load current project"
```
Comment (iwahbe) [inline sdk/go/common/workspace/creds.go:42]: Could this be a structured error that we could use with `errors.Is` instead of a string comparison?

**Pull request closed without merging**